vault/logical_system_raft.go

```diff
@@ -335,6 +382,121 @@ func (b *SystemBackend) handleStorageRaftSnapshotRead() framework.OperationFunc
+				"voters":                       state.Voters,
+				"non_voters":                   state.NonVoters,
+			},
+		}, nil
+	}
+}
+
+func (b *SystemBackend) handleStorageRaftAutopilotConfigRead() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		if !ok {
+			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
+		}
+
+		config := raftStorage.AutopilotConfig()
+		if config == nil {
+			return nil, nil
+		}
+
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"cleanup_dead_servers":      config.CleanupDeadServers,
+				"last_contact_threshold":    config.LastContactThreshold.String(),
+				"max_trailing_logs":         config.MaxTrailingLogs,
+				"min_quorum":                config.MinQuorum,
+				"server_stabilization_time": config.ServerStabilizationTime.String(),
+			},
+		}, nil
+	}
+}
+
+func (b *SystemBackend) handleStorageRaftAutopilotConfigUpdate() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		if !ok {
+			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
+		}
+
+		// Get the config present in the backend
+		config := raftStorage.AutopilotConfig()
+
+		// Mutate the clone of the actual config to avoid setting values in
+		// failure cases.
+		configClone := config.Clone()
+
+		persist := false
+		cleanupDeadServers, ok := d.GetOk("cleanup_dead_servers")
+		if ok {
+			configClone.CleanupDeadServers = cleanupDeadServers.(bool)
+			persist = true
+		}
+		lastContactThreshold, ok := d.GetOk("last_contact_threshold")
+		if ok {
+			configClone.LastContactThreshold = time.Duration(lastContactThreshold.(int)) * time.Second
+			persist = true
+		}
+		maxTrailingLogs, ok := d.GetOk("max_trailing_logs")
+		if ok {
+			configClone.MaxTrailingLogs = uint64(maxTrailingLogs.(int))
+			persist = true
+		}
+		minQuorum, ok := d.GetOk("min_quorum")
+		if ok {
+			configClone.MinQuorum = uint(minQuorum.(int))
+			persist = true
+		}
+		serverStabilizationTime, ok := d.GetOk("server_stabilization_time")
+		if ok {
+			configClone.ServerStabilizationTime = time.Duration(serverStabilizationTime.(int)) * time.Second
+			persist = true
+		}
+
+		if configClone.CleanupDeadServers && configClone.MinQuorum < 3 {
+			return logical.ErrorResponse("min_quorum must be set when cleanup_dead_servers is set and it should at least be 3"), logical.ErrInvalidRequest
+		}
+
+		if persist {
+			entry, err := logical.StorageEntryJSON(raftAutopilotConfigurationStoragePath, configClone)
+			if err != nil {
+				return nil, err
+			}
+			if err := b.Core.barrier.Put(ctx, entry); err != nil {
+				return nil, err
+			}
+		}
+
+		raftStorage.SetAutopilotConfig(configClone)
```

Comment:
Worth logging the new config (since we log it when autopilot starts)?

Reply:
Yep. Updated.